### ISSUE TITLE
CsvParser.stream()

### ISSUE THREAD

**Issue by PawelPacholek** [uniVocity/univocity-parsers#78]
It would be very useful, if CsvParser can be converted to Stream (java 8). One could then easily use it for functional coding (with lazy evaluation). For example instead of this code:

```
private void parseCsv(CsvParser parser) {
  String[] row;
  while ((row = parser.parseNext()) != null)
    if (isAppropriate(row))
       doSomethingWith(row);
}
```

one could have that one:

```
private void parseCsv(CsvParser parser) {
  parser.stream()
    .filter(this::isAppropriate)
    .forEach(this::doSomethingWith);
}
```

**Comment by jbax** [comment]
Will be implemented on a future 3.0.0 version, which will have support for Java 8

**Comment by lgoldstein** [comment]
Several things to remember though

- A Java 8 _Stream_ does not expect any `IOException`(s) so need to figure out how to propagate such exception that may occur during the `while ((row=parser.nextRow()) != null)` loop.
- How to implement single result terminators - e.g., `findFirst/findAny` - that need to stop the parsing loop
- Parallel streams and I/O require extra care - synchronized access while reading a row ?

**Comment by pharmazone** [comment]
what is estimations date for 3.0.0 release?

**Comment by jbax** [comment]
We will start working on it after our HTML parser is released (by the end
on June).

Version 3 will possibly come out by the end of July, but keep in mind we
don't have fixed release dates as things can change.

**Comment by pharmazone** [comment]
awesome, thanx!

**Comment by carbotaniuman** [comment]
I'll throw my 2 cents into the issue. I think that following the Files.lines does it for exceptions is a good bet. Same for synchronization.

I think that keeping the CsvParserBuilder before the .stream() is a good call, and I think this would make the library feel a lot more modern.

**Comment by carbotaniuman** [comment]
@jbax I tried forking your code but I get 19 test failures as of this commit (https://github.com/carbotaniuman/univocity-parsers/commit/f2f866b9b943e2cd4996968472134c754e11a482) when running test. Is there a config error on my side?

**Comment by jbax** [comment]
Are you on windows? I'm on vacation with a mac and this change might have
revealed test cases that need updating or will fail under windows because
of line ending differences.

Should work on mac/linux and that's why I didn't spot any error (yet)

**Comment by carbotaniuman** [comment]
Yes I am. Here's the complete list of test failures for when you get back (what fun).

https://gist.github.com/carbotaniuman/3e4a9cb584025c82f71749d833ea5b5f

**Comment by jbax** [comment]
@camerondavison fixed. The tests failed because they use `\n` as the line ending and on windows the default is `\r\n`.

**Comment by carbotaniuman** [comment]
@jbax I still get errors on the tests as of 356ce43.

https://gist.github.com/carbotaniuman/3e4a9cb584025c82f71749d833ea5b5f

**Comment by carbotaniuman** [comment]
@jbax I think adopting the Files.readLines() method of wrapping a spliterator with StreamSupport is a good call. Is progress going to be on 3.0.0 now?

**Comment by gurpal2000** [comment]
Just a side - I found this small snippet on scalaz-stream which attempts to create a stream.

```
def toStream[Record](parser: CsvParser) = 
  Stream.unfold(parser) { p => 
     Option(p.parseNextRecord()).map((_, p)) }
```

I was trying to create a Java equivalent using unfold from the protonpack library, but can't quite get it right. Below works for a finite (ie. using limit()).

```
  private void Stream<Record> toStream(CsvParser csvParser) {
    return StreamUtils
      .unfold(csvParser, f -> Optional.of(csvParser))
      .map(x -> x.parseNextRecord());
  }
```

**Comment by skyhirider** [comment]
Found this open issue when searching for OpenCsv alternatives. Looking for a parser with stream and seems OpenCsv supports it already so will have to use that library instead.
